File: merge_sort/merge_sort.go
package main

import (
    "fmt"
    "math/rand"
    "time"
)

// Merge sort function. Sort the given array.
// :param array []int: the array to be sorted
func merge_sort(array []int) {
    var i, j, k , half int
    var left, right []int
    size := len(array)
    if size > 1 {
        // divide the array in two
        half = size / 2
        left = make([]int, half)
        right = make([]int, size - half)
        copy(left, array[0:half])
        copy(right, array[half:])
        // Call recursivelly
        merge_sort(left)
        merge_sort(right)
        // Merge back
        i = 0
        j = 0
        k = 0
        for i < len(left) && j < len(right) {
            if left[i] < right[j] {
                array[k] = left[i]
                i++
            } else {
                array[k] = right[j]
                j++
            }
            k++
        }
        // fill the rest
        for i < len(left) {
            array[k] = left[i]
            i++
            k++
        }
        for j < len(right) {
            array[k] = right[j]
            j++
            k++
        }
    }
}

// The main function
func main () {
    rand.Seed(time.Now().UnixNano())
    arr := make([]int, 10);
    for i:=0; i<len(arr); i++ {
        arr[i] = rand.Intn(99)
    }
    fmt.Println(arr)
    merge_sort(arr)
    fmt.Println(arr)
}

File: merge_sort/merge_sort.js




// Merge sort function. Sort the given array.
// :param arr Array(): the array to be sorted
function merge_sort(arr) {
    if (arr.length > 1) {
        var half = Math.floor(arr.length / 2);
        var left = arr.slice(0, half);
        var right = arr.slice(half, arr.length);
        merge_sort(left);
        merge_sort(right);
        var i = 0;
        var j = 0;
        var k = 0;
        // merge back again
        while (i < left.length && j < right.length) {
            if (left[i] < right[j]) {
                arr[k] = left[i];
                i++;
            } else {
                arr[k] = right[j];
                j++;
            }
            k++;
        }
        // finish
        while (i < left.length) {
            arr[k] = left[i];
            i++;
            k++;
        }
        while (j < right.length) {
            arr[k] = right[j];
            j++;
            k++;
        }
    }
}



var a = [9, 7, 8, 6, 5, 3, 4, 1, 2];

console.log(a);
merge_sort(a);
console.log(a);
